mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala

```diff
@@ -566,8 +582,8 @@ class Word2VecModel private[spark] (
     wordList.zip(cosVec)
       .toSeq
       .sortBy(-_._2)
-      .take(num + 1)
-      .tail
+      .filter(tup => wordOpt.map(w => !w.equals(tup._1)).getOrElse(true) && tup._2 != 1.0d)
```

Comment:
It should be equivalent to what you suggest. Something like...

```
val topn1 = wordList.zip(cosVec).toSeq.sortBy(-_._2).take(num + 1)
if (wordOpt.isDefined) {
  topn1.filter(tup => !wordOpt.get.equals(tup._1)).take(num)
} else {
  topn1.tail
}
```

OK that's a bit different than what you suggested but does that help a bit?